R/exercise.R

```diff
@@ -217,49 +258,111 @@ evaluate_exercise <- function(exercise, envir, evaluate_global_setup = FALSE) {
+
+  # construct a global setup chunk to set knitr options
+  knitr_setup_header <- "```{r learnr-setup, include=FALSE}"
+  # hack the pager function so that we can print help with custom pager function
   # http://stackoverflow.com/questions/24146843/including-r-help-in-knitr-output
-  pager <- function(files, header, title, delete.file) {
-    all.str <- do.call("c",lapply(files,readLines))
-    cat(all.str,sep="\n")
+  knitr_setup_body <- paste0(
+    c("options(pager=function(files, header, title, delete.file) {
+        all.str <- do.call(\"c\",lapply(files,readLines))
+        cat(all.str,sep=\"\\n\")
+      })",
+      "knitr::opts_chunk$set(echo = FALSE)",
+      "knitr::opts_chunk$set(comment = NA)",
+      "knitr::opts_chunk$set(error = FALSE)"),
+    collapse = "\n"
+  )
+  knitr_setup_footer <- "\n```"
+  knitr_setup_rmd <- paste0(c(knitr_setup_header, knitr_setup_body, knitr_setup_footer), collapse = "\n")
+
+  # helper function that processes a list of raw setup chunks and
+  # returns a single character vector of knitr chunks for an Rmd file
+  get_chunk_rmds <- function(chunks) {
+    if (is.null(chunks)) return(NULL)
+    setup_rmds <- vapply(chunks, character(1), FUN = function(chunk_info) {
+        # construct the knitr Rmd for exercise and its setup chunks
+        # handle exercise chunk differently from setup chunks
+        if (identical(chunk_info$label, exercise$label)) {
+          # grab exercise code
+          code <- exercise$code
+          # manually set exercise relevant options, disable other options
+          static_opts <- list(include = TRUE,
+                              eval = TRUE,
+                              echo = FALSE,
+                              tutorial = NULL,
+                              cache = FALSE,
+                              child = NULL
+          )
+          # construct a character of all of the options
+          opts <- unpack_options(
+            preserved_opts = chunk_info$opts,
+            inherited_opts = exercise$options,
+            static_opts = static_opts
+          )
+        } else {
+          # grab setup code
+          code <- chunk_info$code
+          # set `include` to false for setup chunks to prevent printing last value
+          static_opts <- list(include = FALSE)
+          # for setup chunk, we don't include any exercise options (inherited_opts)
+          opts <- unpack_options(
+            preserved_opts = chunk_info$opts,
+            inherited_opts = list(),
+            static_opts = static_opts
+          )
+        }
+        # if there's an engine option it's non-R code
+        engine <- chunk_info$engine
+        # we quote the label to ensure that it is treated as a label and not a symbol for instance
+        label_opts <- paste0(c(engine, dput_to_string(chunk_info$label), opts), collapse = ", ")
+        chunk_header <- paste0("```{", label_opts, "}\n")
+        chunk_footer <- "\n```"
+        paste0(chunk_header, code, chunk_footer)
```

Comment:
```suggestion
        paste0(
          paste0("```{", label_opts, "}"), 
          paste0(code, collapse = "\n"), 
          "```", 
          collapse = "\n"
        )
```

Reply:
For some reason the collapse of newline does not work between the header and the code part, so I'll keep as is. Might be good to be explicit about the different parts too like header, code footer etc.

Reply:
The outer paste0 should have a sep not a collapse parameter

Reply:
In this case sep won't work either, we just have to wrap the 3 characters in a single `c()` and use `collapse="\n"`:

```r
paste0(
  c(
    paste0("```{", label_opts, "}"),
    code,
    "```"
    ),
  collapse = "\n"
)
```

Does that look okay?

Reply:
Was thinking of
```
        paste0(
          paste0("```{", label_opts, "}"), 
          paste0(code, collapse = "\n"), 
          "```", 
          sep = "\n"
        )
```

Reply:
Lovely! That works.